Provide box.info.synchro interface (#3140)
tarantool/doc: doc/reference/reference_lua/box_info/synchro.rst

..  _box_info_synchro:

================================================================================
box.info.synchro
================================================================================

..  module:: box.info

..  data:: synchro

    Since version :doc:`2.8.1 </release/2.8.1>`.
    Show the current state of synchronous replication.

    In :ref:`synchronous replication <repl_sync>`, a transaction is considered committed only after achieving
    the required quorum number.
    While the transactions are collecting confirmations from remote nodes, these transactions are waiting in the queue.

    The following information is provided:

    *   ``queue``:

        -   ``owner`` -- ID of the replica that owns the synchronous transaction queue.
            Once an owner instance appears, all other instances become read-only.

>             If the ``owner`` field is ``0``, then every instance is writeable,

Reviewer: if owner is 0, then every instance **may** be writeable.

Author: Fixed

            but they can't create any synchronous transactions.
            To claim or reclaim the queue, use :ref:`box.ctl.promote() <box_ctl-promote>` on the instance that you want
            to promote.
            With elections enabled, the instance runs ``box.ctl.promote()`` automatically after winning the elections.

>             Since version :doc:`2.10.0 </release/2.10.0>`.

Reviewer: I think it would be more clear to put the "Since version ..." labels at the beginning of the relevant paragraph.
Otherwise it looks like "Since version 2.10.0" corresponds only to the last sentence.

Author: I agree. Moved version label to the beginning of the section.

        -   ``term`` (since version :doc:`2.10.0 </release/2.10.0>`) -- current queue term.
            It contains the term of the last ``PROMOTE`` request.
            Usually, it is equal to :ref:`box.info.election.term <box_info_election>`.

>             However, the queue term value may be less than the corresponding one in the election term.

Reviewer 2: `the corresponding one in` - does this part add meaning? The sentence looks fine without it.

```suggestion
            However, the queue term value may be less than the election term.
```

>             It can happen when a new round of elections started, but no instance has been promoted yet.

Reviewer 2: ```suggestion
            It can happen when a new round of elections has started, but no instance has been promoted yet.
```

>         -   ``len`` -- current number of entries that are waiting in the queue.

Reviewer 2: ```suggestion
        -   ``len`` -- the number of entries that are currently waiting in the queue.
```

        -   ``len`` -- current number of entries that are waiting in the queue.

>         -   ``busy`` -- the boolean value is set to ``true`` if there is a synchronous transaction in progress.
>             Until the active transaction is complete, any other incoming synchronous transactions will be delayed.
>             Since version :doc:`2.10.0 </release/2.10.0>`.

Reviewer: ```suggestion
        -   ``busy`` -- the boolean value is set to ``true`` when the instance is processing or writing some system request modifying the queue. For example, PROMOTE, CONFIRM or ROLLBACK.
            Until the request is complete, any other incoming synchronous transactions and system requests will be delayed.
            Since version :doc:`2.10.0 </release/2.10.0>`.
```

>     *   ``quorum`` -- evaluated value of the

Reviewer 2: "evaluated value" sound awkward. I also don't get what this means. Can we rewrite this clearer?

        tarantool> box.info.synchro
        ---
        - queue:
            owner: 1
            term: 2
            len: 0
            busy: false
          quorum: 1
        ...

    **Example 2:**

    First, set a quorum number and a timeout for synchronous replication using the following command:

    ..  code-block:: tarantoolsession

        tarantool> box.cfg{
                 > replication_synchro_quorum=2,
                 > replication_synchro_timeout=1000
                 > }

    Next, check the current state of synchronous replication:

    ..  code-block:: tarantoolsession

        tarantool> box.info.synchro
        ---
        - queue:
            owner: 1
            term: 2
            len: 0
            busy: false
          quorum: 2
        ...

    Create a space called ``sync`` and enable synchronous replication on this space.
    Then, create an index.

    ..  code-block:: tarantoolsession

        tarantool> s = box.schema.space.create("sync", {is_sync=true})
        tarantool> _ = s:create_index('pk')

    After that, use ``box.ctl.promote()`` function to claim a queue:

    ..  code-block:: tarantoolsession

        tarantool> box.ctl.promote()

>     Next, use the ``replace`` command:

Reviewer 2: "Use the replace command" is correct but doesn't reflect a user's action. Let's rewrite this to smth like "Insert some tuples/data ...", "perform data maniupulations" or whatever you like :)

    ..  code-block:: tarantoolsession

        tarantool> require('fiber').new(function() box.space.sync:replace{1} end)
        ---
        - status: suspended
          name: lua
          id: 119
        ...
        tarantool> require('fiber').new(function() box.space.sync:replace{1} end)
        ---
        - status: suspended
          name: lua
          id: 120
        ...
        tarantool> require('fiber').new(function() box.space.sync:replace{1} end)
        ---
        - status: suspended
          name: lua
          id: 121
        ...

>     If you use the ``box.info.synchro`` command again,

Reviewer 2: ```suggestion
    If you call the ``box.info.synchro`` command again,
```
